File: Java/acmicpc-java/acmicpc_19843.java
// 19843 수면 패턴

import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        Map<String, Integer> map = new HashMap<>();
        map.put("Mon", 1);
        map.put("Tue", 2);
        map.put("Wed", 3);
        map.put("Thu", 4);
        map.put("Fri", 5);

        int t = sc.nextInt();
        int n = sc.nextInt();
        sc.nextLine();

        int total = 0, time = 0;
        for (int i = 0; i < n; i++) {
            String[] arr = sc.nextLine().split(" ");            
            time = Integer.valueOf(arr[3]) - Integer.valueOf(arr[1]);

            if (!arr[0].equals(arr[2])) {
                time = time + 24 + (map.get(arr[2]) - map.get(arr[0]) - 1) * 24;
            }

            total = total + time;
        }

        if (total >= t) {
            System.out.println(0);
        }
        else if (t - total > 48) {
            System.out.println(-1);
        }
        else {
            System.out.println(t - total);
        }
    }
}

File: Python/acmicpc-py/15000_20000_py/acmicpc_19843.py
# 19843 수면 패턴

import sys
input = sys.stdin.readline

dic = {
    "Mon" : 1,
    "Tue" : 2,
    "Wed" : 3,
    "Thu" : 4,
    "Fri" : 5
}

t, n = map(int, input().split())
total = 0

for _ in range(n):
    arr = input().split()
    
    time = int(arr[3]) - int(arr[1]) 
    if arr[0] != arr[2]:
        hour = time + 24
        dayTime = (dic[arr[2]] - dic[arr[0]] - 1) * 24
        time = hour + dayTime
    total = total + time

if total >= t:
    print(0)
elif total > 48:
    print(-1)
else:
    print(t - total)
